### PR TITLE
Add cache on kubelet podlist

### PR DESCRIPTION
**Pull request by mfpierre**
### What does this PR do?

Add a caching layer to the `/pods` endpoint of the kubelet API.

### Motivation

Avoid stressing the kubelet + performance

### CODE REVIEW

File: pkg/util/kubernetes/kubelet/kubelet.go [hunk 1]
```diff
@@ -122,20 +129,40 @@ func (ku *KubeUtil) GetHostname() (string, error) {
...
+
+	cached, hit := cache.Cache.Get(podListCacheKey)
+	if hit {
+		pods, ok = cached.(PodList)
+		if !ok {
+			log.Errorf("Invalid pod list cache format, forcing a cache miss")
+			hit = false
+		}
 	}
+	if !hit {
```
Comment (JulienBalestra) [inline pkg/util/kubernetes/kubelet/kubelet.go]: We can return fast here:
```go
if hit {
return pods.Items, nil
}
```

File: pkg/util/docker/docker_util.go [hunk 2]
```diff
@@ -425,12 +425,15 @@ func (d *DockerUtil) Inspect(id string, withSize bool) (types.ContainerJSON, err
...
+	if hit {
 		container, ok = cached.(types.ContainerJSON)
 		if !ok {
-			log.Errorf("invalid cache format, forcing a cache miss")
+			log.Errorf("Invalid inspect cache format, forcing a cache miss")
+			hit = false
 		}
-	} else {
+	}
+	if !hit {
```
Comment (JulienBalestra) [inline pkg/util/docker/docker_util.go]: I think we can return if hit is true.

**Review by JulienBalestra — CHANGES_REQUESTED** [review]
Two nits, thanks for the ResetCache !

**Review by JulienBalestra — DISMISSED** [review]

**Comment by xvello** [comment]
Looks great, the caching/reset logic could use a dedicated unit test though.
You can test we didn't send a request by using a select on the chan:
```go
select {
	case <-kubelet.Requests:
		assert.FailNow("podlist request should have been cached")
	default:
}
```

**Review by JulienBalestra — APPROVED** [review]
lgtm, :+1: for the cleanup

**Pull request merged**